### ISSUE TITLE
Position Refinement with multiple processes

### ISSUE THREAD

**Issue by wfeschen** [ptycho/ptypy#146]
Hi,

I want to implement a position refinement algorithm [1] to the existing DM-engine. My implementation works fine when I run Ptypy in a single process.
`mpiexec -n 1 python scripy.py`
But when I run the same script on multiple processes e.g.
`mpiexec -n 2 python script.py`
python freezes at the first iteration where I update the positions. 

In the position refinement algorithm I generate random shifts for each position and calculate the Fourier Error for the new positions. If the Fourier Error decreased, I keep the shifted position otherwise I will keep the old position. I did this by editing the engine_iterate method in the DM class. 

```
    def engine_iterate(self, num=1):
        """
        Compute `num` iterations.
        """
        to = 0.
        tf = 0.

        for it in range(num):
            t1 = time.time()

            # Fourier update
            error_dct = self.fourier_update()

            t2 = time.time()
            tf += t2 - t1

            # Overlap update
            self.overlap_update()

            t3 = time.time()
            to += t3 - t2

            # count up
            self.curiter +=1

        logger.info('Time spent in Fourier update: %.2f' % tf)
        logger.info('Time spent in Overlap update: %.2f' % to)
        error = parallel.gather_dict(error_dct)

        """
        Position refinement
        """
        if self.p.pos_ref_stop > self.curiter >= self.p.pos_ref_start and self.curiter % self.p.pos_ref_cycle == 0 \
                and self.p.pos_ref:
            self.pos_ref()           

        """
        Pos refinement end
        """

        return error
```

In the pos_ref() function, which I call at the end of the engine_iterate function I cycle through all pods and try to refine the positions.

```
    def pos_ref(self):
        pod_names = self.pods.keys()
        pod_names.sort()

        for pod_name in pod_names:
            pod = self.pods[pod_name]
            if pod.active:
                # calculate random positions
                ......
                pod.ob_view.coord = shifted_position
                self.ob.reformat()
                error = self.get_fourier_error(pod_name)
                .....
               
```

I always call the `self.ob.reformat()` method when I change the coordinates which is somehow critical, because without calling this line python will not freeze. Is this even the intended way to update the pods to the new coordinates? Or is there a better way?

I also figured out that python is not freezing when I do not use the `if pod.active:` statement, I am not sure, but I think I need this for mpi? I am thankful for all help.

Best
Wilhelm

 
[1] https://www.sciencedirect.com/science/article/pii/S0304399112001222

**Comment by pierrethibault** [comment]
Hi Wilhelm,

Great, I hope you submit a PR with this nice addition! 

You probably have a *deadlock* because `ob.reformat` contains MPI calls and you put it in the `if pod.active` block, which is entered by a single node and skipped by all others. As far as I can tell there is no way to run your code in an MPI-efficient way because all nodes need to be informed of changes in storages. Essentially you will have no choice but to do this part sequentially.

So the simplest is to remove the `if pod.active` block and make sure that the rest of your code is MPI-compatible. This means that the random coordinate has to be the same for all nodes (either your random number generator uses the same seed on all nodes *or* you generate the coordinate in one node and broadcast it to the others). You also want to make sure that `get_fourier_error` is also MPI aware. The implementation could start with a `if pod.active`, and finish with a broadcast of the value computed by the active node.

So something like this:
```
    def pos_ref(self):
        pod_names = self.pods.keys()
        pod_names.sort()

        for pod_name in pod_names:
            pod = self.pods[pod_name]
            # calculate random position
            if parallel.master:
                position_shift = np.random.normal(scale = [sigma, sigma])
            position_shift = parallel.bcast(position_shift)
            pod.ob_view.coord += position_shift
            self.ob.reformat()
            error = self.get_fourier_error(pod_name)
            .....

    def get_fourier_error(self, pod_name):
        pod = self.pods[pod_name]
        error = 0.
        if pod.active:
            .....
            error = ....
        error = parallel.allreduce(error)
        return error
```
Hope that helps,

Pierre

**Comment by wfeschen** [comment]
Hello Pierre,

thanks for your response. I am sorry that my answer took me so long. I think that I understood the problem and implemented a version which somehow works in parallel. I had to implement a work-around where I temporary copy the object container (`self.ob`). I create a View to this container where I change the coordinates. 

So my code works now the following way. After an iteration of the ML or DM engine I call the `pos_ref()` function. Here I save a temporary version (`self.temp_ob`) of the `self.ob` container which I use later in the `single_pos_ref()` function to calculate the shifted version of the object. This copy is necessary because otherwise the shape of `self.ob.data` will change for different processes and will cause some troubles. The new positions are calculated in the `single_pos_ref()` function.

```
def pos_ref(obj):
    """
    :param obj: Self of the calling engine class.
    """
    ... 

    # Only used for calculating the shifted pos
    obj.temp_ob = obj.ob.copy()

    for i, di_view_name in enumerate(obj.di.views):
        di_view = obj.di.views[di_view_name]
        pos_num = int(di_view.ID[1:])

        if i == 0:
            # create accessrule
            obj.ar = DEFAULT_ACCESSRULE.copy()
            obj.ar.psize = obj.temp_ob.storages["S00G00"].psize
            obj.ar.shape = obj.shape

        if di_view.active:
            new_coords[pos_num, :] = single_pos_ref(obj, di_view_name)

    new_coords = parallel.allreduce(new_coords)

    for di_view_name in obj.di.views:
        di_view = obj.di.views[di_view_name]
        pos_num = int(di_view.ID[1:])
        if new_coords[pos_num, 0] != 0 and new_coords[pos_num, 1] != 0:
            log(4, "Position: " + str(pos_num))
            log(4, "Old coordinate: " + str(di_view.pod.ob_view.coord))
            log(4, "New coordinate: " + str(new_coords[pos_num, :]))
            di_view.pod.ob_view.coord = new_coords[pos_num, :]

    # Change the coordinates of the object
    obj.ob.reformat()

    t_pos_f = time.time()
    log(4, "Pos ref time: " + str(t_pos_f - t_pos_s))

  ...

```

In the single_pos_ref() function I create a new accessrule which I use to create a new View to the object with shifted coordinates. It can happen that the View on the object is cut (which means that the `ob_view_temp` array will have a different shape than the diffraction pattern. In this case I add zeros to the array). 

```
def single_pos_ref(obj, di_view_name):
    """
    Obj
    Refines a single position.
    """
    di_view = obj.di.views[di_view_name]
    number_rand_shifts = obj.p.number_rand_shifts           # should be a multiple of 4
    dcoords = np.zeros((number_rand_shifts + 1, 2)) - 1.
    pxl_size_obj = obj.ob.S.values()[0].psize[0]            # Pixel size in the object plane
    amplitude = obj.p.amplitude                             # still has to be replaced by parameter value
    end = obj.p.pos_ref_stop
    start = obj.p.pos_ref_start
    max_shift_allowed = obj.p.max_shift_allowed
    it = obj.curiter                                        # current iteration
    max_shift_dist = amplitude * (end - it) / (end - start) + pxl_size_obj/2.
    shape = obj.shape

    delta = np.zeros((number_rand_shifts, 2))               # coordinate shift
    errors = np.zeros(number_rand_shifts)                   # calculated error for the shifted position
    coord = np.copy(di_view.pod.ob_view.coord)

    obj.ar.coord = coord
    obj.ar.storageID = obj.temp_ob.storages["S00G00"].ID

    # Create temporal object view that can be shifted without reformatting
    ob_view_temp = View(obj.temp_ob, accessrule=obj.ar)
    dcoords[0, :] = ob_view_temp.dcoord

    # This can be optimized by saving existing iteration fourier error...
    error_inital = get_fourier_error(di_view, ob_view_temp.data)

    for i in range(number_rand_shifts):

        delta_y = np.random.uniform(0, max_shift_dist)
        delta_x = np.random.uniform(0, max_shift_dist)

        if i % 4 == 1:
            delta_y *= -1
            delta_x *= -1
        elif i % 4 == 2:
            delta_x *= -1
        elif i % 4 == 3:
            delta_y *= -1

        delta[i, 0] = delta_y
        delta[i, 1] = delta_x

        rand_coord = [coord[0] + delta_y, coord[1] + delta_x]
        norm = np.linalg.norm(rand_coord - obj.initial_pos[int(di_view_name[1:]), :])

        if norm > max_shift_allowed:
            # Positions drifted too far, skip this position
            log(4, "New position is too far away!!!", parallel=True)
            errors[i] = error_inital + 1.
            continue

        obj.ar.coord = rand_coord
        ob_view_temp = View(obj.temp_ob, accessrule=obj.ar)
        dcoord = ob_view_temp.dcoord                            # coordinate in pixel

        .... some code for zero padding ...
        errors[i] = get_fourier_error(di_view, new_obj)
     

   if np.min(errors) < error_inital:
        # if a better coordinate is found
        log(4, "New coordinate with smaller Fourier Error found!", parallel=True)
        arg = np.argmin(errors)
        new_coordinate = np.array([coord[0] + delta[arg, 0], coord[1] + delta[arg, 1]])

    else:
        new_coordinate = (0, 0)
    return new_coordinate

     
```
The following parameters have to be defined in the starting script.

Parameters:
**pos_ref_start**: At what iteration the position correction should start.
**pos_ref_stop**: At what iteration the position correction should end.
**pos_ref_cycle**: How often will be searched for better positions. This means for example every second iteration (for pos_ref_cycle = 2).
**max_shift_allowed**: The distance that positions are allowed to drift away from the original position.
**number_rand_shifts**: Number of random shifts which are calculate for each position in every position correction iteration. Right now this should be always a mutliple of 4 since for every direction 1 shift is calculated. There is probably a better way to do this.
**amplitude**: In what range around the original position should be the new randomly calculated position.

To quantify the position improvement I distorted a scan map (mean distance to the original coordiantes was 11px) and the position refinement algorithm was able to refine the positions to a residual shift of ~3px.

How should I proceed? What needs to be done to fulfill the requirements of a PR? Are there any comments on the code? 

Best
Wilhelm

PS:
I had some problems with the views of the shifted coordinates. If the coordinates were shifted too far I had to zero pad the object. I encountered the following unexpected behaviour. If the coordinates of the view exceed the grid in negative direction the data of the view will be empty. On the other hand if the coordinates exceed the grid in positive direction the data of the view will be cut.

```
    # Some example code to demonstrate the unexpected behaviour of views.
    c = Container(data_type="real")
    s = c.new_storage(shape=(1, 10, 10))

    y, x = s.grids()
    s.fill(x + y)
    fig = u.plot_storage(s, 0)

    # adding a view
    ar = DEFAULT_ACCESSRULE.copy()
    ar.coord = (0, 0)
    ar.shape = (4, 4)
    v = View(c, ID=None, accessrule=ar)
    print(v.data.shape)
    # this will print: (4, 4)

    ar = DEFAULT_ACCESSRULE.copy()
    ar.coord = (-2, 0)
    ar.shape = (4, 4)
    v = View(c, ID=None, accessrule=ar)
    print(v.data.shape)
    # this will print: (0, 4)

    ar = DEFAULT_ACCESSRULE.copy()
    ar.coord = (2, 0)
    ar.shape = (4, 4)
    v = View(c, ID=None, accessrule=ar)
    print(v.data.shape)
    # this will print: (2, 4)
```

**Comment by pierrethibault** [comment]
Hi Wilhelm,

Regarding the zero-padding problem you mention last, this happens because you need to call `c.reformat()` before accessing the data. If you want to make sure that your storage doesn't shrink when you reformat, you can set `s.padonly = True`.

Otherwise this looks like a nice addition to ptypy. Would you consider submitting a pull request? Ideally this should be done against the new release that we will roll out in a couple of days.

**Comment by wfeschen** [comment]
Hi Pierre,

Thanks for your response, I really would like to contribute to this project, but I am not sure If I can make it in the next few days. I will submit a pull request after some proper testing of the code, I hope that is fine.

**Comment by pierrethibault** [comment]
Hi Wilhelm,
No worries, there is no rush. Glad to know you are willing to push your work. I suggest that you wait until the next release, which should happen very soon, and submit your PR onto master.

**Comment by daurer** [comment]
Hi Wilhelm, 
Hi Pierre,

I needed some position refinement for my project and I found this interesting thread. So I pulled this code from your fork and made a new 'position_correction' branch in the main repo. I've also done a few changes:

- I integrated a position update into the BaseEngine, such that this feature can be used by any engine. 
- I created a new PosRefine class in a separate posref.py module encapsulating all the position refinement business (basically a transformation of your utils_pos_corr.py)
- I've integrated the position update into DM (just after the overlap update)

Although the code seemed to work fine, it unfortunately did not bring the results that I was hoping for (probably more because of the specifics of my project). But I am happy that I found this nice code and I hope it will be useful for other ptypy users. 

For a full integration of this feature, there are still a few things to be done. I only tested and used it with DM. So it still needs to be properly integrated for the ML engine. I kept the ML_pos_corr.py file for now, but it should be removed once its content has been merged into ML.py

Related to the MPI/View access behaviour discussed above. I like the current approach of creating a temporary copy of the object in each MPI process, because it avoids MPI communication issues. And it seems to be fine to do this, because refining the position a given view can be done independently of all the other positions. What I noticed is that creating the object copy drops the connection to all the associated views and therefore ```reformat()```  (which requires access to all the views) cannot be used here to fix the zero-padding problem. Your current solution to deal with the zero-padding yourself seems to work, but I guess its still worthwhile thinking of how this could be implemented using more of ptypy's already existing tools. 

Sorry for the long text, just wanted to share my current insights after working with this code for a few days. 

@pierrethibault should I submit a PR based on this branch?

Benedikt
